### DISCUSSION TITLE
Consider supporting other data stores

### DISCUSSION THREAD

**Discussion by micahhausler** [tinkerbell/tink#510] [Ideas]
Tinkerbell relies on Postgres as its datastore for hardware, template, and workflow data (and a few other workflow tables). Postgres is a fantastic database, but managing high-availability with graceful failover is not trivial. 

Using an alternative data store that better supports failure would better help operators to have higher availability and not require downtime for upgrades or failover. There are several great open source options like [etcd](https://etcd.io/) or [consul](https://www.consul.io/) that could be discussed in a later proposal. Both projects support watches that could support [streaming workflow data](https://github.com/tinkerbell/tink/blob/a56e5cf/grpc-server/tinkerbell.go#L32-L49) to workers or other clients

I realize this is not a small change, and this isn't a formal technical proposal, but just a tracking discussion for the concept.

**Comment by mmlb** [comment]
Sounds good to me, I've been toying with using sqlite also for ease of dev/test. sqlite is easier as its still SQL, but getting the storage interface correct would be easier if we have a kv store in mind too.

**Comment by mddeff** [comment]
Long time lurker here; I think flexibility in back-end datastores would be awesome.  On the thread of using KV stores ala consul; would there be any advantage to integrating something like `vault` to support storage of secrets?  I could see that being useful for storing something like BMC passwords/credentials for PBnJ (unless I mis-understand PBnJ's workflow).

**Reply by mmlb** [reply]
Hmm yeah secrets is something sorely missing. I don't have a good/well formed idea on how we'd want to go about it too much. Enabling vault integration would definitely be something I'd like to see.
